contracts/TopShot.cdc

```diff
@@ -654,10 +665,95 @@ pub contract TopShot: NonFungibleToken {
+                    let editionList: [MetadataViews.Edition] = [editionInfo]
+                    return MetadataViews.Editions(
+                        editionList
+                    )
+                case Type<MetadataViews.Serial>():
+                    return MetadataViews.Serial(
+                        UInt64(self.data.serialNumber)
+                    )
+                case Type<MetadataViews.Royalties>():
+                    return MetadataViews.Royalties(
+                        royalties: [
+                            // No royalty for Moments yet defined for 3rd party marketplaces
+                        ]
+                    )
+                case Type<MetadataViews.ExternalURL>():
+                    return MetadataViews.ExternalURL(self.getMomentURL())
+                case Type<MetadataViews.NFTCollectionData>():
+                    return MetadataViews.NFTCollectionData(
+                        storagePath: /storage/MomentCollection,
+                        publicPath: /public/MomentCollection,
+                        providerPath: /private/MomentCollection,
+                        publicCollection: Type<&TopShot.Collection{TopShot.MomentCollectionPublic}>(),
+                        publicLinkedType: Type<&TopShot.Collection{TopShot.MomentCollectionPublic,NonFungibleToken.Receiver,NonFungibleToken.CollectionPublic,MetadataViews.ResolverCollection}>(),
+                        providerLinkedType: Type<&TopShot.Collection{NonFungibleToken.Provider}>(),
+                        createEmptyCollectionFunction: (fun (): @NonFungibleToken.Collection {
+                            return <-TopShot.createEmptyCollection()
+                        })
+                    )
+                case Type<MetadataViews.NFTCollectionDisplay>():
+                    let media = MetadataViews.Media(
+                        file: MetadataViews.HTTPFile(
+                            url: "https://nbatopshot.com/static/img/top-shot-logo-horizontal-white.svg"
+                        ),
+                        mediaType: "image/svg+xml"
+                    )
+                    return MetadataViews.NFTCollectionDisplay(
+                        name: "NBA TopShot",
+                        description: "NBA Top Shot is your chance to own, sell, and trade official digital collectibles of the NBA and WNBA's greatest plays and players",
+                        externalURL: MetadataViews.ExternalURL("https://nbatopshot.com"),
+                        squareImage: media,
+                        bannerImage: media,
+                        socials: {}
+                    )
+                case Type<MetadataViews.Traits>():
+                    let traitDictionary: {String: AnyStruct} = {
+                        "Date Of Moment": TopShot.getPlayMetaDataByField(playID: self.data.playID, field: "DateOfMoment"),
```

Comment:
Any chance we can find a way to get Tier in here as well? Happy to help contribute to that if it's possible. That seems like the last piece to get all TS data on-chain

Reply:
Sure thing! Added you to the forked repo if you'd like to add to the PR. 
Currently off chain, moments have a 'tier'..  in all current use cases, all moments for a 'play' will share the tier so it should be safe to add to play metadata for now. 
Previously moments inherited their tier from the _set_ but now have the option to override when they are minted. 
Backfilling tier for all existing plays should be possible (would need a job for this). Also, we'd likely want all newly added moments to add their tier (perhaps a new dictionary of moment id to tier would be appropriate?)

Reply:
> Sure thing! Added you to the forked repo if you'd like to add to the PR.

🙏 If we can get some kind of approval for how Tier could be added, I'm happy to help out there for sure

> This makes the assumption that the following play metadata will be added ....

Since you are already adding data to plays, can Tier just be added as well in one go? It looks like for newly created plays that would be fairly simple, the metadata supplied for the play should just include `Tier`:
https://github.com/dapperlabs/nba-smart-contracts/blob/b6a8ab1611010900c87d3feacadddcc9508d2bbd/contracts/TopShot.cdc#L774

The risky part probably is about adding data to existing plays just in terms of allowing that kind of operation (other pieces of data also need to be added either way), but the method could be quite simple:

```
pub fun addTierToPlay(tier: String, playID: UInt64) {
  TopShot.playDatas[newID]!.metadata["Tier"] = tier
}
```

Then it's just a matter of backfilling which is up to Dapper/TS folks.

Reply:
Going to defer on adding a rarity trait for the immediate PR (Agree this still seems valuable! But we're going to try to avoid backfilling play data on chain for building image URLs, in part because  we'd like to avoid building in a path that allows us to edit metadata of existing plays)
In a followup PR would like to explore consider adding a new map that would contain the rarity of each moment that won't require us to modify the existing play metadata and have the appropriate granularity of being per moment instead of per play.